projects/slap_manipulation/src/slap_manipulation/dataloaders/peract_loader.py

```diff
@@ -0,0 +1,582 @@
+            gripper_pose = torch.FloatTensor(data["global_proprio"][:, :-1])
+            new_data.update(
+                {
+                    "trans_action_indices": trans_action_indices,
+                    "rot_grip_action_indices": grip_rot_action_indices,
+                    "ignore_collisions": ignore_colls,
+                    "gripper_pose": gripper_pose,
+                    "gripper_states": gripper_states,
+                    "attention_coords": attention_coords,
+                }
+            )
+
+            # TODO make sure you update the keys used in peract_model.update() method
+            # preprocess and normalize obs
+            rgb, pcd = self._preprocess_inputs(data)
+            new_data["rgb"] = rgb
+            new_data["xyz"] = pcd
+            new_data["action"] = gripper_states
+
+            if self.is_action_valid(new_data):
+                return new_data
+            else:
+                tries += 1
+
+        return new_data
+
+    def _norm_rgb(self, x):
+        return (x.float() / 1.0) * 2.0 - 1.0
+
+    def _preprocess_inputs(self, sample):
+        rgb = sample["rgb"]
+        pcd = sample["xyz"]
+
+        rgb = self._norm_rgb(rgb)
+
+        # not required as we are not using the baselines below peract
+        # obs.append(
+        #     [rgb, pcd]
+        # )  # obs contains both rgb and pointcloud (used in ARM for other baselines)
+        return rgb, pcd
+
+    def is_action_valid(self, sample):
+        """
+        simple method which compares ee-keyframe position with the scene bounds
+        returns False if action is outside the scene bounds
+        """
+        bounds = np.array(self.scene_bounds)
+        if (sample["ee_keyframe_pos"].numpy() > bounds[..., 3:]).any() or (
+            sample["ee_keyframe_pos"].numpy() < bounds[..., :3]
+        ).any():
+            return False
+        return True
+
+    # discretize translation, rotation, gripper open, and ignore collision actions
+    def _get_action(
+        self,
+        sample: dict,
+        rotation_resolution: int,
+        idx: int = -1,
+        max_keypoints=6,
+    ):
+        if idx == -1:
+            quat = sample["ee_keyframe_ori_quat"]
+            attention_coordinate = sample["ee_keyframe_pos"]
+            grip = float(sample["proprio"][-1])
+            D = len(self.voxel_sizes)
+            gripper_width = sample["gripper_width_array"]
+        else:
+            quat = sample["ee_keyframe_ori_quat"][idx]
+            attention_coordinate = sample["ee_keyframe_pos"][idx]
+            grip = float(sample["all_proprio"][idx, -1])
+            # gripper_state = sample["target_gripper_state"][idx]
+            # grip = float(sample["proprio"][idx][-1])
+            gripper_width = sample["gripper_width_array"][idx]
+            time_index = float((2.0 * idx) / max_keypoints)
+        quat = utils.normalize_quaternion(quat)
+        if quat[-1] < 0:
+            quat = -quat
```

Comment:
huh??

Reply:
It's what PerAct does to ensure all quaternions map to a unique bin when they discretize.